### PR TITLE
Introducing `StepNarrow` trait

### PR DESCRIPTION
**Pull request by taikiy**
Taking a baby step with the `Step` optimization.

New: `trait StepNarrow` which defines `fn narrow(&self, step: &S) -> Self;`
Rename:
* `struct Step` -> `struct GenericStep`
* `trait SubStep` -> `trait Step`

Upcoming next is a new `trait Gate` and make `Context` generic over `G: Gate`.

### CODE REVIEW

File: src/protocol/step.rs [hunk 1]
```diff
@@ -0,0 +1,6 @@
+use super::Step;
```
Comment (martinthomson) [inline src/protocol/step.rs:1]: Want to move all the step related structs/traits here?

File: src/test_fixture/world.rs [hunk 2]
```diff
@@ -150,7 +150,7 @@ impl TestWorld {
     }
 
     #[must_use]
-    pub fn execution_step(execution: usize) -> impl Substep {
+    pub fn execution_step(execution: usize) -> impl Step {
```
Comment (martinthomson) [inline src/test_fixture/world.rs:153]: I just realized that this is going to hurt quite a bit when we stop using `GenericStep`.  I think that we're OK with that; we can just not use this feature for non-generic steps, but @akoshelev might not realize his dream of running tests with non-generic steps as a result.  We rely on this to make steps diverse.

File: src/protocol/mod.rs [hunk 3]
```diff
@@ -75,23 +77,22 @@ impl Substep for str {}
     derive(serde::Deserialize),
     serde(from = "&str")
 )]
-pub struct Step {
+pub struct GenericStep {
```
Comment (martinthomson) [inline src/protocol/mod.rs:80]: GenericGate?  Or does that make it too hard to deal with (comments and such).

**Review by martinthomson — APPROVED** [review]

File: src/protocol/mod.rs (hunk 3 above)
Comment (akoshelev) [reply src/protocol/mod.rs:80]: DescriptiveStep? StringBackedStep? not sure what is the best name here and what would be the appropriate name for "efficient" step backed by an int

File: src/test_fixture/world.rs (hunk 2 above)
Comment (akoshelev) [reply src/test_fixture/world.rs:153]: Looked into it as well - it seems we can make it work if we consume the world to run a test (that shouldn't require narrowing)

```rust
pub trait Runner {
    async fn oneshot<I, A, O, H, R>(self, input: I, helper_fn: H) -> [O; 3];
    ...
}
```

agree with @martinthomson that we don't need this feature for all tests, likely we only need to run IPA with non generic step

**Review by akoshelev — APPROVED** [review]

File: src/protocol/step.rs (hunk 1 above)
Comment (taikiy) [reply src/protocol/step.rs:1]: Yes, will move them in the next PR.

File: src/protocol/mod.rs (hunk 3 above)
Comment (taikiy) [reply src/protocol/mod.rs:80]: strings
* DescriptiveStep
* StringStep
* StringBackedStep
* VerboseStep
* TextualStep

integers
* IntStep
* NumericStep
* CompactStep
* MemoryEfficientStep

Comment (akoshelev) [reply src/protocol/mod.rs:80]: I vote for `DescriptiveStep` and `CompactStep`

**Pull request merged**